src/main/scala/za/co/absa/commons/version/Version.scala

```diff
@@ -36,4 +38,20 @@ object Version extends SimpleVersionImpl with SemVer20Impl {
     def semver(args: Any*): Version = Version.asSemVer(sc.s(args: _*))
   }
 
+  implicit class VersionExtensionMethods(val v: Version) extends AnyVal {
+
+    def asString: String = {
+      @tailrec def loop(sb: StringBuilder, delim: String, comps: List[Component]): StringBuilder = comps match {
+        case Nil => sb
+        case EmptyComponent :: cs => loop(sb, ".", cs)
+        case NumericComponent(x) :: cs => loop(sb.append(delim).append(x), ".", cs)
+        case StringComponent(s) :: cs => loop(sb.append(delim).append(s), ".", cs)
+        case PreReleaseComponent(identifiers@_*) :: cs => loop(sb, "-", identifiers.toList ++ cs)
+        case BuildMetadataComponent(identifiers@_*) :: cs => loop(sb, "+", identifiers.toList ++ cs)
+      }
+
+      loop(StringBuilder.newBuilder, "", v.components.toList).result()
+    }
```

Comment:
Why not put the method directly into Version class?

Reply:
Because `Version` is designed as a case class, and case classes override `toSting()`. I didn't want to alter standard case class behavior. Also, `toString()` method in Java is often abused. A good practice is to only use `toString()` for logging/debugging purposes, but _never_ for business logic.
In the current case, rendering a version string can be seen as a process opposite to the parsing. Thus I put it into an extension method, somewhat close to the parser implementation.